Delete build folder after successful conan test (#2160)

niosHD opened the pull request:
Additionally, by default a temp folder is used as build directory and
a new command line option can be used to specify a folder if needed.

memsharded commented on conans/client/cmd/test.py:
```diff
@@ -12,31 +12,37 @@ def __init__(self, manager, user_io):
         """
         base_folder = os.path.dirname(conanfile_abs_path)
-        build_folder = self._build_folder(profile, base_folder)
+        delete_build_folder = build_folder is None
```
As you can see Linux tests are passing, Windows tests are failing. This could be due to the fact that Windows hold handles to files open, more time than necessary, so it is not possible to remove the build folder just after building and running the package tests.

niosHD replied:
I already fixed this issue. The problem was that, on windows, deleting the current working directory is not possible.

niosHD commented:
I know that this change breaks the current flow. However, I think that now is the last chance to fix the (in my opinion) incorrect default. As you mentioned in #2160, the majority of the use cases don't require the build directory. Using a temporary folder is the correct default setting in this case.

Additionally, the current behavior with the sha1 subfolder is also not really ideal for those who require the build artifacts of the test package (outside of the conanfile.py). Using the new flag, the users that actually care about the build folder can specify that they care and where the folder should be located.

I am really certain that the current default is incorrect and that we should fix it as soon as possible. However, in the end you guys have to decide. If you feel strongly that the current behavior is correct I respect that and will update this PR.

niosHD commented:
I am fine with switching to `--test-build-folder`. However, I still want to have the possibility to use a temporary folder outside of the source tree by default.

Would it be acceptable to add an option into `conan.conf` which permits to opt-in the changed default behavior?

memsharded commented:
> Would it be acceptable to add an option into conan.conf which permits to opt-in the changed default behavior?

In fact I think this would be a better approach, no need for a new command-line argument, as this is likely something that users might want as a general behavior, not per command invocation (and if need to change per invocation, it is just an env-var, so they can do it).

niosHD commented:
I rebased the PR and updated it such that the default behavior can be configured via `general.temp_test_folder` in `conan.conf` and via the `CONAN_TEMP_TEST_FOLDER` environment variable. The default is of course to stick with the current behavior. Note that I also slightly changed the added behavior and delete the build folder now unconditionally of the build result. Finally, I also kept the `--test-build-folder` command line arguments for `conan create` and `conan test` because I think they are useful.

memsharded commented:
Thanks for doing the rebase. However, it still seems there are some conflicts, could you please have a look?

Regarding the changes, they look good. If I understand correctly the code, it is only doing the automatic deletion if the conan.conf or env-var is defining to use temporary storage, for both the default or user-custom explicit folder, it won't remove it. I'd say it is perfect. Also, good testing.

Also, regarding the create & test command line arguments, I don't oppose. I will approve it, waiting for conflict to be solved, and final check from @lasote

niosHD commented:
Sure, looks like I screwed up the rebase yesterday, should be fixed now.

Correct, the build directory for the test package is only deleted automatically when the use of the temporary build directory is enabled via conan.conf or env-var and no manually test build folder has been specified by the user.

Great, thank you James for the fast and positive feedback.

lasote commented:
Could you solve the conflict? it is just a couple of imports. It can be merged later.

niosHD commented:
Rebased, but the CI is failing now. Looks like a Jenkins problem though. What do you think @lasote?
